File: packages/picoipc/examples/cpp/producer.cpp
#include "smh_q/ring.hpp"

#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <thread>

int main(int argc, char** argv) {
  const std::string name = (argc > 1) ? argv[1] : "smh_q_demo";
  const int count = (argc > 2) ? std::atoi(argv[2]) : 10;
  const int delay_ms = (argc > 3) ? std::atoi(argv[3]) : 100;

  smh_q::Ring::Config cfg;
  cfg.name = name;
  cfg.slot_count = 16;
  cfg.slot_size = 128;

  smh_q::Ring ring(cfg, true);
  std::printf("producer: name=%s count=%d delay_ms=%d\n", name.c_str(), count, delay_ms);

  for (int i = 0; i < count; ++i) {
  retry:
    char buf[64];
    const int n = std::snprintf(buf, sizeof(buf), "msg-%04d", i);
    while (!ring.try_publish(std::span<const std::byte>(
        reinterpret_cast<const std::byte*>(buf), static_cast<std::size_t>(n)))) {
      std::this_thread::sleep_for(std::chrono::milliseconds(1));
      goto retry;
    }
    std::printf("published: %s (write_seq=%u)\n", buf, ring.write_seq());
    std::this_thread::sleep_for(std::chrono::milliseconds(delay_ms));
  }

  std::puts("producer done");
  return 0;
}

File: packages/picoipc/python/examples/producer.py
#!/usr/bin/env python3
"""Producer: writes numbered messages into the ring."""

import argparse
import os
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from picoipc import Ring


def main() -> int:
    p = argparse.ArgumentParser(description="picoipc SPSC producer")
    p.add_argument("name", nargs="?", default="smh_q_demo")
    p.add_argument("count", nargs="?", type=int, default=10)
    p.add_argument("delay_ms", nargs="?", type=int, default=100)
    args = p.parse_args()

    ring = Ring(args.name, create=True, slot_count=16, slot_size=128)
    print(f"producer: name={args.name} count={args.count} delay_ms={args.delay_ms}")

    for i in range(args.count):
        msg = f"msg-{i:04d}".encode()
        while not ring.try_publish(msg):
            time.sleep(0.001)
        print(f"published: {msg.decode()} (write_seq={ring.write_seq()})")
        time.sleep(args.delay_ms / 1000.0)

    print("producer done")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
